base/intset.jl

```diff
@@ -1,307 +1,236 @@
-symdiff(s1::IntSet, s2::IntSet) =
-    (s1.limit >= s2.limit ? symdiff!(copy(s1), s2) : symdiff!(copy(s2), s1))
-
-function empty!(s::IntSet)
-    s.bits[:] = 0
-    return s
+intersect(s1::IntSet, s2::IntSet) = intersect!(copy(s1), s2)
+function intersect!(s1::IntSet, s2::IntSet)
+    l = length(s2.bits)
+    if     !s1.inverse & !s2.inverse;  _resize0!(s1.bits, l);         map!(&, s1.bits, s1.bits, s2.bits)
+    elseif  s1.inverse & !s2.inverse;  _resize0!(s1.bits, l);         map!(<, s1.bits, s1.bits, s2.bits); s1.inverse = false
+    elseif !s1.inverse &  s2.inverse;  e = _matchlength!(s1.bits, l); map!(>, s1.bits, s1.bits, s2.bits); append!(s1.bits, e)
+    else #= s1.inverse &  s2.inverse=# e = _matchlength!(s1.bits, l); map!(|, s1.bits, s1.bits, s2.bits); append!(s1.bits, e)
+    end
+    s1
 end
 
-function symdiff!(s::IntSet, n::Integer)
-    if n >= s.limit
-        lim = Int(n + dim(n,2))
-        sizehint!(s, lim)
-    elseif n < 0
-        throw(ArgumentError("IntSet elements cannot be negative"))
+setdiff(s::IntSet, ns) = setdiff!(copy(s), ns)
+setdiff!(s::IntSet, ns) = (for n in ns; _delete!(s, n); end; s)
+function setdiff!(s1::IntSet, s2::IntSet)
+    l = length(s2.bits)
+    if     !s1.inverse & !s2.inverse;  e = _matchlength!(s1.bits, l); map!(>, s1.bits, s1.bits, s2.bits); append!(s1.bits, e)
+    elseif  s1.inverse & !s2.inverse;  e = _matchlength!(s1.bits, l); map!(|, s1.bits, s1.bits, s2.bits); append!(s1.bits, e)
+    elseif !s1.inverse &  s2.inverse;  _resize0!(s1.bits, l);         map!(&, s1.bits, s1.bits, s2.bits)
+    else #= s1.inverse &  s2.inverse=# _resize0!(s1.bits, l);         map!(<, s1.bits, s1.bits, s2.bits); s1.inverse = false
     end
-    s.bits[n>>5 + 1] $= (UInt32(1)<<(n&31))
-    return s
+    s1
 end
 
-function symdiff!(s::IntSet, ns)
-   for n in ns
-       symdiff!(s, n)
-   end
-   return s
+symdiff(s::IntSet, ns) = symdiff!(copy(s), ns)
+symdiff!(s::IntSet, ns) = (for n in ns; symdiff!(s, n); end; s)
+function symdiff!(s::IntSet, n::Integer)
+    0 <= n < typemax(Int) || throw(ArgumentError(_intset_bounds_err_msg))
+    val = (n in s) $ !s.inverse
+    _setint!(s, n, val)
+    s
 end
-
-function copy!(to::IntSet, from::IntSet)
-    empty!(to)
-    union!(to, from)
+function symdiff!(s1::IntSet, s2::IntSet)
+    e = _matchlength!(s1.bits, length(s2.bits))
+    map!($, s1.bits, s1.bits, s2.bits)
+    s2.inverse && (s1.inverse = !s1.inverse)
+    append!(s1.bits, e)
+    s1
 end
 
 function in(n::Integer, s::IntSet)
-    if n >= s.limit
-        # max IntSet length is typemax(Int), so highest possible element is
-        # typemax(Int)-1
-        s.fill1s && n >= 0 && n < typemax(Int)
-    elseif n < 0
-        return false
+    idx = n+1
+    if 1 <= idx <= length(s.bits)
+        unsafe_getindex(s.bits, idx) != s.inverse
     else
-        (s.bits[n>>5 + 1] & (UInt32(1)<<(n&31))) != 0
+        ifelse((idx <= 0) | (idx > typemax(Int)), false, s.inverse)
     end
 end
```

Comment:
That would probably be more clear, yes, but this works just fine as it is.

The result of the last expression in a function body is used as the return value (if not explicitly specified).  In this case, the last expression is the `if` block.

The result of an `if` block is similarly determined by the last expression of the traversed branch (or `nothing` if no branches were traversed).